torch/_higher_order_ops/scan.py

```diff
@@ -357,9 +456,176 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
     return generic_scan(combine_fn, init, xs, dim, reverse, additional_inputs)
 
 
-scan_op.py_impl(DispatchKey.Autograd)(
-    autograd_not_implemented(scan_op, deferred_error=True)
-)
+class ScanAutogradOp(torch.autograd.Function):
+    @staticmethod
+    def forward(
+        ctx,
+        fw_graph,
+        joint_graph,
+        dim,
+        reverse,
+        num_leaves_init,
+        num_leaves_additional_inputs,
+        *ops,
+    ):
+        ops_list = list(ops)
+        init = ops_list[:num_leaves_init]
+        xs = ops_list[num_leaves_init : len(ops_list) - num_leaves_additional_inputs]
+        additional_inputs = ops_list[len(ops_list) - num_leaves_additional_inputs :]
+
+        ctx._joint_graph = joint_graph
+        ctx._dim = dim
+        ctx._reverse = reverse
+        ctx._num_leaves_init = num_leaves_init
+        ctx._num_leaves_xs = len(xs)
+
+        with torch._C._AutoDispatchBelowAutograd():
+            carry, carries_outs = _extract_carry_and_out(
+                scan_op(fw_graph, init, xs, dim, reverse, additional_inputs),
+                num_leaves_init,
+            )
+
+            # Collect the carries for each time step from the outs
+            # and save them for the backward path
+            carries = carries_outs[:num_leaves_init]
+            outs = carries_outs[num_leaves_init:]
+            ctx.save_for_backward(*(init + xs + carries + additional_inputs))
+            ctx._num_leaves_ys = len(outs)
+
+            return (*carry, *outs)
+
+    @staticmethod
+    def backward(ctx, *flat_grads):
+        r"""
+        This function computes the gradients of the scan operation.
+        It does so by constructing using an additional scan operator with the gradients
+
+        Args:
+            flat_grads (torch.Tensor): The tensor of upstream gradients, or a nested pytree of tensors.
+
+        Example::
+
+            The ``fw_graph`` f(.,.), used in the forward function, is the operator used during the scan. For example
+            def f(x: torch.Tensor, y: torch.Tensor):
```

Comment:
this example is a bit boring: maybe change to next_carry = y = x * y?

Reply:
I rewrote the example for the multiply function